File: legacy/server-node/src/discover.js
/**
 * Descubrimiento de routers candidatos en la LAN:
 *
 *  1. Subred: se deriva de la puerta de enlace por defecto (`ip route`),
 *     asumiendo /24 (LAN doméstica).
 *  2. Barrido TCP del puerto 22 en los 254 hosts con concurrencia limitada
 *     (NO se usa ping: necesita cap_net_raw, no disponible en LXC ni con el
 *     hardening del servicio).
 *  3. A los vivos se les prueba la firma OpenWrt: POST http://<ip>/ubus
 *     (cualquier respuesta JSON-RPC, incluso "access denied", delata uhttpd
 *     con ubus ⇒ router OpenWrt/GL.iNet).
 *  4. A los candidatos se les prueba SSH con la clave propia (BatchMode):
 *       - OK → `authorized: true` + modelo (ubus system board)
 *       - Permission denied → `authorized: false` (falta autorizar la clave)
 *
 * El resultado se cachea 60 s (el barrido tarda varios segundos).
 */
import net from 'node:net'
import http from 'node:http'
import https from 'node:https'
import { execFile } from 'node:child_process'
import { detectGatewayIp, listRouters } from './routerstore.js'
import { sshBaseArgs } from './sshkey.js'

const CACHE_MS = 60_000
const TCP_TIMEOUT_MS = 800
const HTTP_TIMEOUT_MS = 1500
const SSH_TIMEOUT_MS = 4000

let cache = { at: 0, subnet: null, results: [] }

function exec(cmd, args, timeoutMs) {
  return new Promise((resolve) => {
    execFile(cmd, args, { timeout: timeoutMs }, (err, stdout) => {
      resolve({ ok: !err, stdout: stdout || '' })
    })
  })
}

/** true si el puerto TCP responde (SYN/ACK) antes del timeout. */
function tcpOpen(host, port, timeoutMs = TCP_TIMEOUT_MS) {
  return new Promise((resolve) => {
    const sock = net.connect({ host, port, timeout: timeoutMs })
    sock.once('connect', () => {
      sock.destroy()
      resolve(true)
    })
    sock.once('timeout', () => {
      sock.destroy()
      resolve(false)
    })
    sock.once('error', () => resolve(false))
  })
}

/** Ejecuta tareas con concurrencia limitada. */
async function pool(items, size, fn) {
  const out = new Array(items.length)
  let i = 0
  await Promise.all(
    Array.from({ length: Math.min(size, items.length) }, async () => {
      while (i < items.length) {
        const idx = i++
        out[idx] = await fn(items[idx])
      }
    }),
  )
  return out
}

/** POST /ubus crudo (http u https con cert ignorado) — devuelve {status, body} o null. */
function postUbus(host, useHttps) {
  return new Promise((resolve) => {
    const body = JSON.stringify({
      jsonrpc: '2.0',
      id: 1,
      method: 'call',
      params: ['0'.repeat(32), 'session', 'login', { username: 'netpulse-probe', password: '' }],
    })
    const mod = useHttps ? https : http
    const req = mod.request(
      {
        host,
        port: useHttps ? 443 : 80,
        path: '/ubus',
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(body) },
        timeout: HTTP_TIMEOUT_MS,
        rejectUnauthorized: false,
      },
      (res) => {
        let data = ''
        res.on('data', (chunk) => {
          data += chunk
          if (data.length > 8192) req.destroy()
        })
        res.on('end', () => resolve({ status: res.statusCode ?? 0, body: data }))
      },
    )
    req.on('timeout', () => {
      req.destroy()
      resolve(null)
    })
    req.on('error', () => resolve(null))
    req.write(body)
    req.end()
  })
}

/**
 * true si el host huele a OpenWrt/GL.iNet:
 *  1. POST /ubus (http, luego https si hay redirect) con firma JSON-RPC.
 *  2. El firmware GL.iNet NO expone /ubus (nginx lo redirige a su UI) →
 *     firma alternativa: el HTML raíz contiene "gl-ui"/"GL.iNet".
 */
async function probeUbus(host) {
  const isUbus = (r) => r && (r.body.includes('jsonrpc') || r.body.includes('ubus_rpc_session'))
  const httpRes = await postUbus(host, false)
  if (isUbus(httpRes)) return true
  if (httpRes && [301, 302, 307, 308].includes(httpRes.status)) {
    if (isUbus(await postUbus(host, true))) return true
    const root = await getRoot(host, true)
    if (root && /gl-ui|GL\.iNet|glinet/i.test(root.body)) return true
  }
  return false
}

/** GET / (http/https, cert ignorado) — devuelve {status, body} o null. */
function getRoot(host, useHttps) {
  return new Promise((resolve) => {
    const mod = useHttps ? https : http
    const req = mod.request(
      { host, port: useHttps ? 443 : 80, path: '/', method: 'GET', timeout: HTTP_TIMEOUT_MS, rejectUnauthorized: false },
      (res) => {
        let data = ''
        res.on('data', (chunk) => {
          data += chunk
          if (data.length > 16384) req.destroy()
        })
        res.on('end', () => resolve({ status: res.statusCode ?? 0, body: data }))
        res.on('error', () => resolve(null))
      },
    )
    req.on('timeout', () => {
      req.destroy()
      resolve(null)
    })
    req.on('error', () => resolve(null))
    req.end()
  })
}

/** Prueba SSH con la clave propia; devuelve { authorized, model }. */
async function probeSsh(host, sshKeyPath) {
  const { ok, stdout } = await exec(
    'ssh',
    [
      ...sshBaseArgs(sshKeyPath),
      '-o', 'ConnectTimeout=2',
      '-o', 'ControlMaster=no',
      `root@${host}`,
      'ubus call system board | jsonfilter -e @.model',
    ],
    SSH_TIMEOUT_MS,
  )
  if (!ok) return { authorized: false, model: null }
  return { authorized: true, model: stdout.trim() || null }
}

/**
 * Escanea la LAN y devuelve candidatos a router OpenWrt.
 * @param {object} db - handle better-sqlite3 (para marcar los ya configurados)
 * @param {string} sshKeyPath
 * @param {boolean} force - ignora la caché
 */
export async function discoverRouters(db, sshKeyPath, force = false) {
  if (!force && Date.now() - cache.at < CACHE_MS) return { ...cache, cached: true }

  const gwIp = await detectGatewayIp()
  if (!gwIp) return { subnet: null, results: [], cached: false, error: 'no_gateway' }
  const prefix = gwIp.split('.').slice(0, 3).join('.')
  const subnet = `${prefix}.0/24`

  // Hosts con SSH a la vista (barrido TCP :22) — se excluye la propia IP
  const selfIps = new Set(
    (await exec('sh', ['-c', 'ip -o -4 addr show | grep -oP "inet \\K[0-9.]+"'], 3000)).stdout
      .split('\n')
      .filter(Boolean),
  )
  const hosts = Array.from({ length: 254 }, (_, i) => `${prefix}.${i + 1}`).filter((h) => !selfIps.has(h))
  const alive = (
    await pool(hosts, 100, async (h) => ((await tcpOpen(h, 22)) ? h : null))
  ).filter(Boolean)

  // De los vivos, los que huelen a OpenWrt (ubus HTTP)
  const configured = new Set(listRouters(db).map((r) => r.host))
  const candidates = (
    await pool(alive, 16, async (h) => {
      if (!(await probeUbus(h))) return null
      const { authorized, model } = await probeSsh(h, sshKeyPath)
      return {
        host: h,
        isGateway: h === gwIp,
        authorized,
        model,
        configured: configured.has(h),
      }
    })
  ).filter(Boolean)

  candidates.sort((a, b) => Number(b.isGateway) - Number(a.isGateway) || a.host.localeCompare(b.host))
  cache = { at: Date.now(), subnet, results: candidates }
  return { ...cache, cached: false }
}

File: server-go/internal/discover/discover.go
// Package discover — descubrimiento de routers candidatos en la LAN
// (paridad src/discover.js, SPEC §7.7):
//  1. Subred /24 derivada del gateway por defecto (`ip route`).
//  2. Barrido TCP del puerto 22 en los 254 hosts (concurrencia 100; NO ping).
//  3. Firma OpenWrt: POST http://<ip>/ubus (cualquier JSON-RPC, incluso
//     "access denied", delata uhttpd+ubus; GL.iNet redirige a su UI → firma
//     gl-ui por https).
//  4. Probe SSH con la clave propia → {authorized, model}.
//
// Caché de 60 s (force=1 la ignora).
package discover

import (
	"bytes"
	"crypto/tls"
	"database/sql"
	"io"
	"net"
	"net/http"
	"os/exec"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/gnacho/netpulse/server-go/internal/routerstore"
	"github.com/gnacho/netpulse/server-go/internal/sshkey"
)

const (
	cacheMS      = 60_000
	tcpTimeout   = 800 * time.Millisecond
	httpTimeout  = 1500 * time.Millisecond
	sshTimeout   = 4 * time.Second
	scanWorkers  = 100
	probeWorkers = 16
)

// Result es un candidato a router.
type Result struct {
	Host       string  `json:"host"`
	IsGateway  bool    `json:"isGateway"`
	Authorized bool    `json:"authorized"`
	Model      *string `json:"model"`
	Configured bool    `json:"configured"`
}

// Response es la respuesta de /api/config/discover.
type Response struct {
	At      int64    `json:"at,omitempty"` // la respuesta no_gateway de Node NO incluye `at`
	Subnet  *string  `json:"subnet"`
	Results []Result `json:"results"`
	Cached  bool     `json:"cached"`
	Error   string   `json:"error,omitempty"`
}

var (
	cacheMu sync.Mutex
	cache   Response
)

// tcpOpen: true si el puerto TCP responde antes del timeout.
func tcpOpen(host string, port int, timeout time.Duration) bool {
	conn, err := net.DialTimeout("tcp", net.JoinHostPort(host, itoa(port)), timeout)
	if err != nil {
		return false
	}
	_ = conn.Close()
	return true
}

func itoa(n int) string {
	var buf [4]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	return string(buf[i:])
}

// pool ejecuta fn sobre items con concurrencia limitada.
func pool[T any](items []T, size int, fn func(T) *T) []*T {
	out := make([]*T, len(items))
	var idx int64
	var mu sync.Mutex
	var wg sync.WaitGroup
	workers := size
	if len(items) < workers {
		workers = len(items)
	}
	for w := 0; w < workers; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for {
				mu.Lock()
				if idx >= int64(len(items)) {
					mu.Unlock()
					return
				}
				i := idx
				idx++
				mu.Unlock()
				out[i] = fn(items[i])
			}
		}()
	}
	wg.Wait()
	res := make([]*T, 0, len(out))
	for _, r := range out {
		if r != nil {
			res = append(res, r)
		}
	}
	return res
}

func httpClient() *http.Client {
	return &http.Client{
		Timeout: httpTimeout,
		Transport: &http.Transport{
			TLSClientConfig:     &tls.Config{InsecureSkipVerify: true}, // firma, no confianza
			DisableKeepAlives:   true,
			MaxConnsPerHost:     1,
			TLSHandshakeTimeout: httpTimeout,
		},
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse // necesitamos el status crudo (301/302)
		},
	}
}

type probeResp struct {
	status int
	body   string
}

var ubusProbeBody = `{"jsonrpc":"2.0","id":1,"method":"call","params":["00000000000000000000000000000000","session","login",{"username":"netpulse-probe","password":""}]}`

// postUbus: POST /ubus crudo (http u https) — {status, body} o nil.
func postUbus(host string, useHTTPS bool) *probeResp {
	scheme := "http"
	port := "80"
	if useHTTPS {
		scheme = "https"
		port = "443"
	}
	req, err := http.NewRequest("POST", scheme+"://"+host+":"+port+"/ubus", bytes.NewReader([]byte(ubusProbeBody)))
	if err != nil {
		return nil
	}
	req.Header.Set("Content-Type", "application/json")
	res, err := httpClient().Do(req)
	if err != nil {
		return nil
	}
	defer res.Body.Close()
	body, _ := io.ReadAll(io.LimitReader(res.Body, 8192))
	return &probeResp{status: res.StatusCode, body: string(body)}
}

// getRoot: GET / (http/https) — {status, body} o nil.
func getRoot(host string, useHTTPS bool) *probeResp {
	scheme := "http"
	port := "80"
	if useHTTPS {
		scheme = "https"
		port = "443"
	}
	res, err := httpClient().Get(scheme + "://" + host + ":" + port + "/")
	if err != nil {
		return nil
	}
	defer res.Body.Close()
	body, _ := io.ReadAll(io.LimitReader(res.Body, 16384))
	return &probeResp{status: res.StatusCode, body: string(body)}
}

var glUIRe = regexp.MustCompile(`(?i)gl-ui|GL\.iNet|glinet`)

// probeUbus: true si el host huele a OpenWrt/GL.iNet.
func probeUbus(host string) bool {
	isUbus := func(r *probeResp) bool {
		return r != nil && (strings.Contains(r.body, "jsonrpc") || strings.Contains(r.body, "ubus_rpc_session"))
	}
	httpRes := postUbus(host, false)
	if isUbus(httpRes) {
		return true
	}
	if httpRes != nil && (httpRes.status == 301 || httpRes.status == 302 || httpRes.status == 307 || httpRes.status == 308) {
		if isUbus(postUbus(host, true)) {
			return true
		}
		if root := getRoot(host, true); root != nil && glUIRe.MatchString(root.body) {
			return true
		}
	}
	return false
}

// probeSsh prueba SSH con la clave propia; devuelve (authorized, model).
func probeSsh(host, keyPath string) (bool, *string) {
	args := append(sshkey.BaseArgs(keyPath),
		"-o", "ConnectTimeout=2",
		"-o", "ControlMaster=no",
		"root@"+host,
		"ubus call system board | jsonfilter -e @.model",
	)
	cmd := exec.Command("ssh", args...)
	type result struct {
		out []byte
		err error
	}
	ch := make(chan result, 1)
	go func() {
		out, err := cmd.Output()
		ch <- result{out, err}
	}()
	select {
	case <-time.After(sshTimeout):
		_ = cmd.Process.Kill()
		return false, nil
	case res := <-ch:
		if res.err != nil {
			return false, nil
		}
		model := strings.TrimSpace(string(res.out))
		if model == "" {
			return true, nil
		}
		return true, &model
	}
}

// selfIPs: IPs propias del servidor (para excluirlas del barrido).
func selfIPs() map[string]bool {
	out, err := exec.Command("sh", "-c", `ip -o -4 addr show | grep -oP "inet \K[0-9.]+"`).Output()
	ips := map[string]bool{}
	if err != nil {
		return ips
	}
	for _, l := range strings.Split(string(out), "\n") {
		if l = strings.TrimSpace(l); l != "" {
			ips[l] = true
		}
	}
	return ips
}

// Routers escanea la LAN y devuelve candidatos (paridad discoverRouters).
// db se usa solo para marcar los ya configurados.
func Routers(db *sql.DB, keyPath string, force bool) Response {
	cacheMu.Lock()
	defer cacheMu.Unlock()
	now := time.Now().UnixMilli()
	if !force && now-cache.At < cacheMS {
		c := cache
		c.Cached = true
		return c
	}

	gwIP := routerstore.DetectGatewayIP()
	if gwIP == "" {
		return Response{Subnet: nil, Results: []Result{}, Cached: false, Error: "no_gateway"}
	}
	parts := strings.Split(gwIP, ".")
	prefix := strings.Join(parts[:3], ".")
	subnet := prefix + ".0/24"

	own := selfIPs()
	hosts := make([]string, 0, 254)
	for i := 1; i <= 254; i++ {
		h := prefix + "." + itoa(i)
		if !own[h] {
			hosts = append(hosts, h)
		}
	}
	alive := pool(hosts, scanWorkers, func(h string) *string {
		if tcpOpen(h, 22, tcpTimeout) {
			return &h
		}
		return nil
	})

	configured := map[string]bool{}
	for _, r := range routerstore.ListRouters(db) {
		configured[r.Host] = true
	}
	type candidate struct{ h string }
	cands := make([]candidate, 0, len(alive))
	for _, h := range alive {
		cands = append(cands, candidate{*h})
	}
	found := pool(cands, probeWorkers, func(c candidate) *candidate {
		if !probeUbus(c.h) {
			return nil
		}
		return &c
	})

	results := make([]Result, 0, len(found))
	for _, c := range found {
		authorized, model := probeSsh(c.h, keyPath)
		results = append(results, Result{
			Host:       c.h,
			IsGateway:  c.h == gwIP,
			Authorized: authorized,
			Model:      model,
			Configured: configured[c.h],
		})
	}
	sort.Slice(results, func(i, j int) bool {
		if results[i].IsGateway != results[j].IsGateway {
			return results[i].IsGateway
		}
		return results[i].Host < results[j].Host
	})

	cache = Response{At: now, Subnet: &subnet, Results: results, Cached: false}
	return cache
}
